README.md
# slidable_button
[![Pub](https://img.shields.io/pub/v/slidable_button.svg?style=flat-square)](https://pub.dartlang.org/packages/slidable_button)
[![support](https://img.shields.io/badge/platform-flutter%7Cflutter%20web-2196f3.svg?style=flat-square)](https://github.com/husainazkas/slidable_button)

A simple flutter plugin for Slidable Button.

<div style="text-align:center"><img src="https://raw.githubusercontent.com/husainazkas/slidable_button/master/screenshot/1.gif" width="300px"/> &nbsp; <img src="https://raw.githubusercontent.com/husainazkas/slidable_button/master/screenshot/2.gif" width="300px"/></div>

<div style="text-align:center"><img src="https://raw.githubusercontent.com/husainazkas/slidable_button/master/screenshot/3.gif" width="300px" /></div>

Note : This plugin is cloned and remade from [swipe_button](https://pub.dev/packages/swipe_button).

## Getting Started

### Add dependency

```yaml
dependencies:
  slidable_button: ^3.0.0+1
```

### Simple to use

Import this library on your code.

```dart
import 'package:slidable_button/slidable_button.dart';
```

Now you can make your slidable button vertically or horizontally by using one of them widget. For horizontal usage example :

```dart
    HorizontalSlidableButton(
      width: MediaQuery.of(context).size.width / 3,
      buttonWidth: 60.0,
      color: Theme.of(context).accentColor.withOpacity(0.5),
      buttonColor: Theme.of(context).primaryColor,
      dismissible: false,
      label: Center(child: Text('Slide Me')),
      child: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text('Left'),
            Text('Right'),
          ],
        ),
      ),
      onChanged: (position) {
        setState(() {
          if (position == SlidableButtonPosition.end) {
            result = 'Button is on the right';
          } else {
            result = 'Button is on the left';
          }
        });
      },
    ),
```

Vertical usage example :

```dart
    VerticalSlidableButton(
      height: MediaQuery.of(context).size.height / 3,
      buttonHeight: 60.0,
      color: Theme.of(context).accentColor.withOpacity(0.5),
      buttonColor: Theme.of(context).primaryColor,
      dismissible: false,
      label: Center(child: Text('Slide Me')),
      child: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text('Top'),
            Text('Bottom'),
          ],
        ),
      ),
      onChanged: (position) {
        setState(() {
          if (position == SlidableButtonPosition.end) {
            result = 'Button is on the bottom';
          } else {
            result = 'Button is on the top';
          }
        });
      },
    ),
```

### Migration Guide

If you already use this package before v3.x.x you may need to migrate your code to `HorizontalSlidableButton` or  `VerticalSlidableButton` for specific direction usage because `SlidableButton` class has been removed start from v3.0.0.

CHANGELOG.md
## 3.0.0+1
* Update CHANGELOG.

## 3.0.0
* Legacy SlidableButton has been removed, use specified direction instead.
* Can update initial button position using setState.
* Button can be stopped at center position.
* Initial button position can set to center even center point is false.

## 2.0.0+1
* Update docs on VerticalSlidableButton.

## 2.0.0
* `SlidableButton` is now deprecate. Please use `Horizontal` or `VerticalSlidableButton` for specific direction usage.
* Added `autoSlide` as new optional param that can control the button to auto slide after release or not.

## 1.3.1
* Prevention multiple trigger `onChanged` when `tristate` is true.

## 1.3.0
* New state `SlidableButtonPosition.sliding` when `tristate` is set to true.

## 1.2.0
* Added new option when the animation slide start after released [PR #4](https://github.com/husainazkas/slidable_button/pull/4).

## 1.1.0
* Added new option to restart button position [PR #2](https://github.com/husainazkas/slidable_button/pull/2).

## 1.0.2
* Added `HitTestBehavior.translucent`.
* Prevention maintenance from deprecated API.

## 1.0.1+2
* Fix `label` not showing when onChanged is null.
* Add example supports for Android, iOS, Web, macOS, and Windows.

## 1.0.1+1
* Update README.

## 1.0.1
* Fix border behind the button.

## 1.0.0+1
* Update README.

## 1.0.0
* Initial release.
* Support null safety.

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:slidable_button_example/main.dart';

void main() {
  testWidgets('Slidable Button Test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const SlidableButtonDemo());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && widget.data!.startsWith('Slide Me'),
      ),
      findsOneWidget,
    );
  });
}


lib/src/vertical_slidable_button.dart
import 'package:flutter/material.dart';

import '../slidable_button.dart';
import 'slidable_button_clipper.dart';
import 'slidable_button_simulation.dart';

class VerticalSlidableButton extends StatefulWidget {
  /// Label of the button.
  final Widget? label;

  /// A widget that is behind the button.
  final Widget? child;

  /// Button color if it disabled.
  ///
  /// [disabledColor] is set to `Colors.grey` by default.
  final Color? disabledColor;

  /// The color of button.
  ///
  /// If null, it will be transparent.
  final Color? buttonColor;

  /// The color of background.
  ///
  /// If null, it will be transparent.
  final Color? color;

  /// Border of area slide (usually called background).
  final BoxBorder? border;

  /// Border Radius for the button and it's child.
  ///
  /// Default value is `const BorderRadius.all(const Radius.circular(60.0))`
  final BorderRadius borderRadius;

  /// The height of this widget (button and it's background).
  ///
  /// Default value is 120.0.
  final double height;

  /// Width of area slide (usually called background).
  ///
  /// Default value is 36.0.
  final double width;

  /// Width of button. If [buttonHeight] is still null and the [label] is not null, this will automatically wrapping [label].
  ///
  /// The minimum size is [height], and the maximum size is three quarters from [width].
  final double? buttonHeight;

  /// It means the effect while and after sliding.
  ///
  /// If `true`, [child] will disappear along with button sliding. Otherwise, it stay visible even the button was slide.
  final bool dismissible;

  /// Initial button position. It can on the left or right.
  final SlidableButtonPosition initialPosition;

  /// The % at which the slide gesture should be considered as completed (0 to 1)
  ///
  /// Default value is 0.5.
  final double completeSlideAt;

  /// Listen to position, is button on the left or right.
  ///
  /// You must set this argument although is null.
  final ValueChanged<SlidableButtonPosition>? onChanged;

  /// Controller for the button while sliding.
  final AnimationController? controller;

  /// Restart animation when the position is opposite to initialPosition
  ///
  /// Default value false
  final bool isRestart;

  /// If true the button's position can be left, right, or sliding. Otherwise only left or right.
  ///
  /// Default value false
  final bool tristate;

  /// Button will auto slide to nearest point after drag released if true, otherwise will not slide.
  ///
  /// Default value true
  final bool autoSlide;

  /// Whether the button can be stopped at center position.
  ///
  /// This will ignore and lock [completeSlideAt] to default value.
  ///
  /// Default value false
  final bool centerPoint;

  /// Creates a [VerticalSlidableButton]
  const VerticalSlidableButton({
    super.key,
    required this.onChanged,
    this.controller,
    this.child,
    this.autoSlide = true,
    this.disabledColor,
    this.buttonColor,
    this.color,
    this.label,
    this.border,
    this.borderRadius = const BorderRadius.all(Radius.circular(60.0)),
    this.initialPosition = SlidableButtonPosition.start,
    double completeSlideAt = 0.5,
    this.height = 120.0,
    this.width = 36.0,
    this.buttonHeight,
    this.dismissible = true,
    this.isRestart = false,
    this.tristate = false,
    this.centerPoint = false,
  }) : completeSlideAt = centerPoint ? 0.5 : completeSlideAt;

  @override
  State<VerticalSlidableButton> createState() => _VerticalSlidableButtonState();
}

class _VerticalSlidableButtonState extends State<VerticalSlidableButton>
    with SingleTickerProviderStateMixin {
  final GlobalKey _containerKey = GlobalKey();
  final GlobalKey _positionedKey = GlobalKey();

  late final AnimationController _controller;
  late Animation<double> _contentAnimation;
  Offset _start = Offset.zero;
  bool _isSliding = false;

  RenderBox? get _positioned =>
      _positionedKey.currentContext!.findRenderObject() as RenderBox?;

  RenderBox? get _container =>
      _containerKey.currentContext!.findRenderObject() as RenderBox?;

  double get _buttonHeight {
    final height = widget.buttonHeight ?? double.minPositive;
    final maxHeight = widget.height * 3 / 4;
    if (height > maxHeight) return maxHeight;
    return height;
  }

  @override
  void initState() {
    super.initState();
    _controller =
        widget.controller ?? AnimationController.unbounded(vsync: this);
    _contentAnimation = Tween<double>(begin: 1.0, end: 0.0)
        .animate(CurvedAnimation(parent: _controller, curve: Curves.easeOut));
    _initialPositionController();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  void didUpdateWidget(covariant VerticalSlidableButton oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (widget.initialPosition != oldWidget.initialPosition) {
      _initialPositionController();
    }
  }

  void _initialPositionController() {
    switch (widget.initialPosition) {
      case SlidableButtonPosition.start:
        _controller.value = 0.0;
        break;
      case SlidableButtonPosition.center:
        _controller.value = 0.5;
        break;
      case SlidableButtonPosition.end:
        _controller.value = 1.0;
        break;
      default:
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: widget.width,
      height: widget.height,
      decoration: BoxDecoration(
        border: widget.border,
        borderRadius: widget.borderRadius,
      ),
      child: Stack(
        key: _containerKey,
        children: <Widget>[
          DecoratedBox(
            decoration: BoxDecoration(
              color: widget.color,
              borderRadius: widget.borderRadius,
            ),
            child: widget.dismissible
                ? ClipRRect(
                    clipper: SlidableButtonClipper(
                      animation: _controller,
                      borderRadius: widget.borderRadius,
                    ),
                    borderRadius: widget.borderRadius,
                    child: SizedBox.expand(
                      child: FadeTransition(
                        opacity: _contentAnimation,
                        child: widget.child,
                      ),
                    ),
                  )
                : SizedBox.expand(child: widget.child),
          ),
          AnimatedBuilder(
            animation: _controller,
            builder: (context, child) => Align(
              alignment: Alignment(0.0, (_controller.value * 2.0) - 1.0),
              child: child,
            ),
            child: Container(
              key: _positionedKey,
              height: _buttonHeight,
              width: widget.width,
              decoration: BoxDecoration(
                borderRadius: widget.borderRadius,
                color: widget.onChanged == null
                    ? widget.disabledColor ?? Colors.grey
                    : widget.buttonColor,
              ),
              child: widget.onChanged == null
                  ? Center(child: widget.label)
                  : GestureDetector(
                      behavior: HitTestBehavior.translucent,
                      onHorizontalDragStart: _onDragStart,
                      onHorizontalDragUpdate: _onDragUpdate,
                      onHorizontalDragEnd: _onDragEnd,
                      child: Center(child: widget.label),
                    ),
            ),
          ),
        ],
      ),
    );
  }

  void _onDragStart(DragStartDetails details) {
    final pos = _positioned!.globalToLocal(details.globalPosition);
    _start = Offset(0.0, pos.dy);
    _controller.stop(canceled: true);
  }

  void _onDragUpdate(DragUpdateDetails details) {
    final pos = _container!.globalToLocal(details.globalPosition) - _start;
    final extent = _container!.size.height - _positioned!.size.height;
    _controller.value = (pos.dy.clamp(0.0, extent) / extent);

    if (widget.tristate && !_isSliding) {
      _isSliding = true;
      _onChanged(SlidableButtonPosition.sliding);
    }
  }

  void _onDragEnd(DragEndDetails details) {
    if (!widget.autoSlide) return _afterDragEnd();

    if (widget.centerPoint) {
      if (_controller.value > 0.25 && _controller.value < 0.75) {
        _controller
            .animateTo(
              0.5,
              duration: const Duration(milliseconds: 500),
              curve: Curves.easeIn,
            )
            .whenComplete(_afterDragEnd);
        return;
      }
    }

    final extent = _container!.size.height - _positioned!.size.height;
    double fractionalVelocity = (details.primaryVelocity! / extent).abs();
    if (fractionalVelocity < 0.5) {
      fractionalVelocity = 0.5;
    }

    double acceleration, velocity;
    if (_controller.value >= widget.completeSlideAt) {
      acceleration = 0.5;
      velocity = fractionalVelocity;
    } else {
      acceleration = -0.5;
      velocity = -fractionalVelocity;
    }

    final simulation = SlidableSimulation(
      acceleration,
      _controller.value,
      1.0,
      velocity,
    );

    _controller.animateWith(simulation).whenComplete(_afterDragEnd);
  }

  void _afterDragEnd() {
    final SlidableButtonPosition position;

    if (_controller.value > 0.5) {
      position = SlidableButtonPosition.end;
    } else {
      if (widget.centerPoint && _controller.value == 0.5) {
        position = SlidableButtonPosition.center;
      } else {
        position = SlidableButtonPosition.start;
      }
    }

    if (widget.isRestart && widget.initialPosition != position) {
      _initialPositionController();
    }

    _isSliding = false;

    _onChanged(position);
  }

  void _onChanged(SlidableButtonPosition position) {
    if (widget.onChanged != null) {
      widget.onChanged!(position);
    }
  }
}


lib/src/horizontal_slidable_button.dart
import 'package:flutter/material.dart';

import '../slidable_button.dart';
import 'slidable_button_clipper.dart';
import 'slidable_button_simulation.dart';

class HorizontalSlidableButton extends StatefulWidget {
  /// Label of the button.
  final Widget? label;

  /// A widget that is behind the button.
  final Widget? child;

  /// Button color if it disabled.
  ///
  /// [disabledColor] is set to `Colors.grey` by default.
  final Color? disabledColor;

  /// The color of button.
  ///
  /// If null, it will be transparent.
  final Color? buttonColor;

  /// The color of background.
  ///
  /// If null, it will be transparent.
  final Color? color;

  /// Border of area slide (usually called background).
  final BoxBorder? border;

  /// Border Radius for the button and it's child.
  ///
  /// Default value is `const BorderRadius.all(const Radius.circular(60.0))`
  final BorderRadius borderRadius;

  /// The height of this widget (button and it's background).
  ///
  /// Default value is 36.0.
  final double height;

  /// Width of area slide (usually called background).
  ///
  /// Default value is 120.0.
  final double width;

  /// Width of button. If [buttonWidth] is still null and the [label] is not null, this will automatically wrapping [label].
  ///
  /// The minimum size is [height], and the maximum size is three quarters from [width].
  final double? buttonWidth;

  /// It means the effect while and after sliding.
  ///
  /// If `true`, [child] will disappear along with button sliding. Otherwise, it stay visible even the button was slide.
  final bool dismissible;

  /// Initial button position. It can on the left or right.
  final SlidableButtonPosition initialPosition;

  /// The % at which the slide gesture should be considered as completed (0 to 1)
  ///
  /// Default value is 0.5.
  final double completeSlideAt;

  /// Listen to position, is button on the left or right.
  ///
  /// You must set this argument although is null.
  final ValueChanged<SlidableButtonPosition>? onChanged;

  /// Controller for the button while sliding.
  final AnimationController? controller;

  /// Restart animation when the position is opposite to initialPosition
  ///
  /// Default value false
  final bool isRestart;

  /// If true the button's position can be left, right, or sliding. Otherwise only left or right.
  ///
  /// Default value false
  final bool tristate;

  /// Button will auto slide to nearest point after drag released if true, otherwise will not slide.
  ///
  /// Default value true
  final bool autoSlide;

  /// Whether the button can be stopped at center position.
  /// 
  /// This will ignore and lock [completeSlideAt] to default value.
  ///
  /// Default value false
  final bool centerPoint;

  /// Creates a [HorizontalSlidableButton]
  const HorizontalSlidableButton({
    super.key,
    required this.onChanged,
    this.controller,
    this.child,
    this.autoSlide = true,
    this.disabledColor,
    this.buttonColor,
    this.color,
    this.label,
    this.border,
    this.borderRadius = const BorderRadius.all(Radius.circular(60.0)),
    this.initialPosition = SlidableButtonPosition.start,
    double completeSlideAt = 0.5,
    this.height = 36.0,
    this.width = 120.0,
    this.buttonWidth,
    this.dismissible = true,
    this.isRestart = false,
    this.tristate = false,
    this.centerPoint = false,
  }) : completeSlideAt = centerPoint ? 0.5 : completeSlideAt;

  @override
  State<HorizontalSlidableButton> createState() =>
      _HorizontalSlidableButtonState();
}

class _HorizontalSlidableButtonState extends State<HorizontalSlidableButton>
    with SingleTickerProviderStateMixin {
  final GlobalKey _containerKey = GlobalKey();
  final GlobalKey _positionedKey = GlobalKey();

  late final AnimationController _controller;
  late Animation<double> _contentAnimation;
  Offset _start = Offset.zero;
  bool _isSliding = false;

  RenderBox? get _positioned =>
      _positionedKey.currentContext!.findRenderObject() as RenderBox?;

  RenderBox? get _container =>
      _containerKey.currentContext!.findRenderObject() as RenderBox?;

  double get _buttonWidth {
    final width = widget.buttonWidth ?? double.minPositive;
    final maxWidth = widget.width * 3 / 4;
    if (width > maxWidth) return maxWidth;
    return width;
  }

  @override
  void initState() {
    super.initState();
    _controller =
        widget.controller ?? AnimationController.unbounded(vsync: this);
    _contentAnimation = Tween<double>(begin: 1.0, end: 0.0)
        .animate(CurvedAnimation(parent: _controller, curve: Curves.easeOut));
    _initialPositionController();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  void didUpdateWidget(covariant HorizontalSlidableButton oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (widget.initialPosition != oldWidget.initialPosition) {
      _initialPositionController();
    }
  }

  void _initialPositionController() {
    switch (widget.initialPosition) {
      case SlidableButtonPosition.start:
        _controller.value = 0.0;
        break;
      case SlidableButtonPosition.center:
        _controller.value = 0.5;
        break;
      case SlidableButtonPosition.end:
        _controller.value = 1.0;
        break;
      default:
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: widget.width,
      height: widget.height,
      decoration: BoxDecoration(
        border: widget.border,
        borderRadius: widget.borderRadius,
      ),
      child: Stack(
        key: _containerKey,
        children: <Widget>[
          DecoratedBox(
            decoration: BoxDecoration(
              color: widget.color,
              borderRadius: widget.borderRadius,
            ),
            child: widget.dismissible
                ? ClipRRect(
                    clipper: SlidableButtonClipper(
                      animation: _controller,
                      borderRadius: widget.borderRadius,
                    ),
                    borderRadius: widget.borderRadius,
                    child: SizedBox.expand(
                      child: FadeTransition(
                        opacity: _contentAnimation,
                        child: widget.child,
                      ),
                    ),
                  )
                : SizedBox.expand(child: widget.child),
          ),
          AnimatedBuilder(
            animation: _controller,
            builder: (context, child) => Align(
              alignment: Alignment((_controller.value * 2.0) - 1.0, 0.0),
              child: child,
            ),
            child: Container(
              key: _positionedKey,
              height: widget.height,
              width: _buttonWidth,
              decoration: BoxDecoration(
                borderRadius: widget.borderRadius,
                color: widget.onChanged == null
                    ? widget.disabledColor ?? Colors.grey
                    : widget.buttonColor,
              ),
              child: widget.onChanged == null
                  ? Center(child: widget.label)
                  : GestureDetector(
                      behavior: HitTestBehavior.translucent,
                      onHorizontalDragStart: _onDragStart,
                      onHorizontalDragUpdate: _onDragUpdate,
                      onHorizontalDragEnd: _onDragEnd,
                      child: Center(child: widget.label),
                    ),
            ),
          ),
        ],
      ),
    );
  }

  void _onDragStart(DragStartDetails details) {
    final pos = _positioned!.globalToLocal(details.globalPosition);
    _start = Offset(pos.dx, 0.0);
    _controller.stop(canceled: true);
  }

  void _onDragUpdate(DragUpdateDetails details) {
    final pos = _container!.globalToLocal(details.globalPosition) - _start;
    final extent = _container!.size.width - _positioned!.size.width;
    _controller.value = (pos.dx.clamp(0.0, extent) / extent);

    if (widget.tristate && !_isSliding) {
      _isSliding = true;
      _onChanged(SlidableButtonPosition.sliding);
    }
  }

  void _onDragEnd(DragEndDetails details) {
    if (!widget.autoSlide) return _afterDragEnd();

    if (widget.centerPoint) {
      if (_controller.value > 0.25 && _controller.value < 0.75) {
        _controller
            .animateTo(
              0.5,
              duration: const Duration(milliseconds: 500),
              curve: Curves.easeIn,
            )
            .whenComplete(_afterDragEnd);
        return;
      }
    }

    final extent = _container!.size.width - _positioned!.size.width;
    double fractionalVelocity = (details.primaryVelocity! / extent).abs();
    if (fractionalVelocity < 0.5) {
      fractionalVelocity = 0.5;
    }

    double acceleration, velocity;
    if (_controller.value >= widget.completeSlideAt) {
      acceleration = 0.5;
      velocity = fractionalVelocity;
    } else {
      acceleration = -0.5;
      velocity = -fractionalVelocity;
    }

    final simulation = SlidableSimulation(
      acceleration,
      _controller.value,
      1.0,
      velocity,
    );

    _controller.animateWith(simulation).whenComplete(_afterDragEnd);
  }

  void _afterDragEnd() {
    final SlidableButtonPosition position;

    if (_controller.value > 0.5) {
      position = SlidableButtonPosition.end;
    } else {
      if (widget.centerPoint && _controller.value == 0.5) {
        position = SlidableButtonPosition.center;
      } else {
        position = SlidableButtonPosition.start;
      }
    }

    if (widget.isRestart && widget.initialPosition != position) {
      _initialPositionController();
    }

    _isSliding = false;

    _onChanged(position);
  }

  void _onChanged(SlidableButtonPosition position) {
    if (widget.onChanged != null) {
      widget.onChanged!(position);
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:slidable_button/slidable_button.dart';

void main() {
  runApp(const MaterialApp(home: SlidableButtonDemo()));
}

class SlidableButtonDemo extends StatefulWidget {
  const SlidableButtonDemo({Key? key}) : super(key: key);

  @override
  State<SlidableButtonDemo> createState() => _SlidableButtonDemoState();
}

class _SlidableButtonDemoState extends State<SlidableButtonDemo> {
  String result = "Let's slide!";

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Slidable Button Demo'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Row(
              children: [
                Expanded(
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      const Text('Slide this button to left or right.'),
                      const SizedBox(height: 16.0),
                      HorizontalSlidableButton(
                        initialPosition: SlidableButtonPosition.start,
                        width: MediaQuery.of(context).size.width / 3,
                        buttonWidth: 60.0,
                        color: Theme.of(context)
                            .colorScheme
                            .secondary
                            .withOpacity(0.5),
                        buttonColor: Theme.of(context).primaryColor,
                        dismissible: false,
                        label: const Center(child: Text('Slide Me')),
                        child: const Padding(
                          padding: EdgeInsets.all(8.0),
                          child: Row(
                            mainAxisAlignment: MainAxisAlignment.spaceBetween,
                            children: [
                              Text('Left'),
                              Text('Right'),
                            ],
                          ),
                        ),
                        onChanged: (position) {
                          setState(() {
                            if (position == SlidableButtonPosition.end) {
                              result = 'Button is at the right';
                            } else {
                              result = 'Button is on the left';
                            }
                          });
                        },
                      ),
                    ],
                  ),
                ),
                const VerticalDivider(),
                Expanded(
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      const Text('Slide this button to top or bottom.'),
                      const SizedBox(height: 16.0),
                      VerticalSlidableButton(
                        initialPosition: SlidableButtonPosition.start,
                        height: MediaQuery.of(context).size.height / 3,
                        buttonHeight: 60.0,
                        color: Theme.of(context)
                            .colorScheme
                            .secondary
                            .withOpacity(0.5),
                        buttonColor: Theme.of(context).primaryColor,
                        dismissible: false,
                        label: const Center(child: Text('Slide Me')),
                        child: const Padding(
                          padding: EdgeInsets.all(8.0),
                          child: Column(
                            mainAxisAlignment: MainAxisAlignment.spaceBetween,
                            children: [
                              Text('Top'),
                              Text('Bottom'),
                            ],
                          ),
                        ),
                        onChanged: (position) {
                          setState(() {
                            if (position == SlidableButtonPosition.end) {
                              result = 'Button is at the bottom';
                            } else {
                              result = 'Button is on the top';
                            }
                          });
                        },
                      ),
                    ],
                  ),
                ),
              ],
            ),
            const SizedBox(height: 16.0),
            Text('Result:\n$result', textAlign: TextAlign.center),
          ],
        ),
      ),
    );
  }
}
